Test FSDP with submodule non-reentrant checkpointing

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89781

With combining FSDP with reentrant checkpointing, the post backward
hook might run twice, and then hit [this
error](https://github.com/pytorch/pytorch/blob/e20ec44544c17d6d3d411f88b870e05043bda731/torch/distributed/fsdp/_runtime_utils.py#L487).
This is because reentrant backward uses nested autograd GraphTasks.
The inner GraphTask is not aware of the outer one and therefore
will flush pending `AccumulateGrad` invocations on exit, which in
turn triggers the post backward hooks registered by FSDP. Later,
the outer GraphTask will trigger that again, leading to the above
error.

PR #89791 relaxes the FSDP training state check, but we still run
into grad value check failures occasionally. Therefore, this PR only
lands the test for non-reentrant test, and we can enable the
reentrant test when the accuracy issues are addressed.

mrshenli committed:
Test FSDP with submodule non-reentrant checkpointing

[ghstack-poisoned]

mrshenli committed:
Update on "Test FSDP with submodule non-reentrant checkpointing"




[ghstack-poisoned]

mrshenli committed:
Update on "Test FSDP with submodule non-reentrant checkpointing"




[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_checkpoint.py:
```diff
@@ -277,6 +279,116 @@ def test_basic_checkpoint_end_to_end(
+        for _, m in model_ac.named_modules():
+            if isinstance(m, CheckpointModule):
+                m.checkpoint = True
+                m.use_reentrant = False
```
why don't we just pass `use_reentrant=False` in L325?

awgu commented on test/distributed/fsdp/test_fsdp_checkpoint.py:
```diff
@@ -277,6 +279,116 @@ def test_basic_checkpoint_end_to_end(
         dist.barrier()
 
 
+class CheckpointModule(nn.Module):
```
nit: Should we add a non-linearity (e.g. ReLU) to this model just to be safe for the gradient correctness check between 'with AC' and 'without AC'?

I do not think it matters for this case, but I just generally avoid the case where the model is equivalent to one affine transformation since I have had silent issues before.

awgu commented on test/distributed/fsdp/test_fsdp_checkpoint.py:
```diff
@@ -277,6 +279,116 @@ def test_basic_checkpoint_end_to_end(
+
+
+class TestFSDPCheckpointSubmodule(FSDPTest):
+
```
I am guessing that trying `use_orig_params=True` can be done in a follow-up. This PR only targets `use_orig_params=False` and `NO_SHARD`?

mrshenli replied:
yep, only targeting a specific case at this moment

awgu commented on test/distributed/fsdp/test_fsdp_checkpoint.py:
```diff
@@ -277,6 +279,116 @@ def test_basic_checkpoint_end_to_end(
+
+        self.assertTrue(model_ac.m1.s1.checkpoint)
+        self.assertTrue(model_ac.m2.s2.checkpoint)
+
```
nit:
```
fsdp_kwargs = {
    "device_id": torch.cuda.current_device(),
    "sharding_strategy": ShardingStrategy.NO_SHARD,
} 
```
and pass `**fsdp_kwargs` to each call to `FSDP()`

awgu reviewed (commented):
Thanks for adding these tests!

At a meta level, I am wondering how we should approach testing more interleavings in a systematic and complete way.

Also, I am not sure if you want to wait for the assert relaxation PR to land and then update `test_checkpoint_submodule_reentrant()` or not.

rohan-varma commented on test/distributed/fsdp/test_fsdp_checkpoint.py:
```diff
@@ -277,6 +279,116 @@ def test_basic_checkpoint_end_to_end(
+        super().__init__()
+        self.l1 = nn.Linear(100, 100)
+        self.m1 = ModelWithCheckpointSubmodule(checkpoint, use_reentrant)
+        self.m2 = ModelWithCheckpointSubmodule(checkpoint, use_reentrant)
```
nit: instead of names `m1` and `m2`, checkpoint_1 and checkpoint_2 might make things clearer for readers.

mrshenli replied:
sure, let me update that

rohan-varma commented on test/distributed/fsdp/test_fsdp_checkpoint.py:
```diff
@@ -277,6 +279,116 @@ def test_basic_checkpoint_end_to_end(
+            sharding_strategy=ShardingStrategy.NO_SHARD,
+        )
+
+        x = torch.randn(2, 100).cuda()
```
nit: `device='cuda'`

rohan-varma commented on test/distributed/fsdp/test_fsdp_checkpoint.py:
```diff
@@ -277,6 +279,116 @@ def test_basic_checkpoint_end_to_end(
+            sharding_strategy=ShardingStrategy.NO_SHARD,
+        )
+
+        x = torch.randn(2, 100).cuda()
```
nit: `device='cuda'`

rohan-varma reviewed (approved):
LGTM, stamping to unblock. Will file a follow up issue to debug why this FSDP + AC structure does not work.

mrshenli committed:
Update on "Test FSDP with submodule non-reentrant checkpointing"


With combining FSDP with reentrant checkpointing, the post backward
hook might run twice, and then hit [this
error](https://github.com/pytorch/pytorch/blob/e20ec44544c17d6d3d411f88b870e05043bda731/torch/distributed/fsdp/_runtime_utils.py#L487).
This is because reentrant backward uses nested autograd GraphTasks.
The inner GraphTask is not aware of the outer one and therefore
will flush pending `AccumulateGrad` invocations on exit, which in
turn triggers the post backward hooks registered by FSDP. Later,
the outer GraphTask will trigger that again, leading to the above
error.

PR #89791 relaxes the FSDP training state check, but we still run
into grad value check failures occasionally. Therefore, this PR only
lands the test for non-reentrant test, and we can enable the
reentrant test when the accuracy issues are addressed.

[ghstack-poisoned]

mrshenli commented:
> Also, I am not sure if you want to wait for the assert relaxation PR to land and then update test_checkpoint_submodule_reentrant() or not.

Updated the PR summary to include that.

mrshenli commented:
@pytorchbot merge -g
